### PR TITLE
wip: refactor pyannote.audio.embedding.approaches

### PR DESCRIPTION
**Pull request by hbredin**
* chore: rename EmbeddingApproach to RepresentationLearning
* feat: inherit from EmbeddingApproach.get_batch_generator
* feat: add per_turn option to all approaches (for aggregation learning)
* BREAKING: remove support for variable length sequences

### CODE REVIEW

**Comment by hbredin** [comment]
cc @juanmc2005 

I spent some time cleaning `pyannote.audio.embedding.approaches`.
This is still very much a work in progress but I'd be love to hear your thoughts about it.

File: pyannote/audio/embedding/approaches/contrastive_loss.py [hunk 1]
```diff
@@ -122,24 +105,29 @@ def batch_loss(self, batch):
...
-            for j in range(i + 1, nbatch):
+        for i in range(batch_size - 1):
+            for j in range(i + 1, batch_size):
                 gt.append(int(y[i] != y[j]))
         gt = torch.Tensor(gt).float().to(self.device_)
 
         # Calculate the losses as described in the paper
         losses = (1 - gt) * torch.pow(dist, 2) + gt * torch.pow(torch.clamp(self.margin_ - dist, min=1e-8), 2)
+
+        # FIXME: why divive by 2?
```
Comment (juanmc2005) [inline pyannote/audio/embedding/approaches/contrastive_loss.py]: The division by 2 comes from equation 4 in the [original paper](http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf)

File: pyannote/audio/embedding/approaches/contrastive_loss.py [hunk 2]
```diff
@@ -28,85 +28,68 @@
...
                        per_epoch: float = None,
-                       label_min_duration=0.):
-
-        super().__init__()
+                       label_min_duration: float = 0.,
+                       # FIXME create a Literal type for metric
+                       # FIXME maybe in pyannote.core.utils.distance
+                       metric: str = 'cosine',
+                       # FIXME homogeneize the meaning of margin parameter
+                       # FIXME it has a different meaning in ArcFace, right?
```
Comment (juanmc2005) [inline pyannote/audio/embedding/approaches/contrastive_loss.py]: yes, the contrastive loss margin specifies how far away the negative pairs should be, whereas the arcface margin specifies how far away we can tolerate positive pairs to be.
About homogenizing the margins, I don't know if it's a good idea because the definition of the margin is inherent to the definition of the loss, like in the example of arcface and contrastive. I think it would be easier when trying to reproduce experiments from papers. For instance, a 0.2 margin in a paper about contrastive loss would mean `ContrastiveLoss(margin=0.2)` in pyannote, otherwise we'd have to calculate which pyannote margin is equivalent to the other one.

File: pyannote/audio/embedding/approaches/contrastive_loss.py [hunk 3]
```diff
@@ -28,85 +28,68 @@
...
+                         per_label=per_label,
+                         per_fold=per_fold,
+                         per_epoch=per_epoch,
+                         label_min_duration=label_min_duration)
 
         self.metric = metric
         self.margin = margin
-
+        # FIXME see above
         self.margin_ = self.margin * self.max_distance
```
Comment (juanmc2005) [inline pyannote/audio/embedding/approaches/contrastive_loss.py]: as above, maybe we should avoid doing this kind of transformations to the margin. What about raising an exception if margin > max_distance ?

**Comment by hbredin** [comment]
Merged it as it is but we should make sure to handle your comments later :)
